### ISSUE TITLE
- temporal stats at team.py and on sparse matrix

### ISSUE THREAD

**Issue by hosseinfani** [fani-lab/OpeNTF#6]
temporal stats seem to be general to all domains. so, they can be moved to the parent class

also, we can add temporal info to the indexes for extracting the stats.

**Comment by hosseinfani** [comment]
Hi @VaghehDashti ,
You need to gradually start this issue to motivate temporal study in your thesis and paper

**Comment by cdcorro** [comment]
@hosseinfani I have created some test functions that I was working on over the past week and was looking into testing them out with the toy data however I can't find the time array data. I am in the dev branch, would you happen to know the file path to the year: index array?

**Comment by hosseinfani** [comment]
@cdcorro 
thank you for the update. 
true. I've run the code on dev on the toy datasets and created the pickle files for you. They're pushed at the following locations:

./data/preprocessed/*/toy.*/*.pkl

**Comment by VaghehDashti** [comment]
> Hi @VaghehDashti , You need to gradually start this issue to motivate temporal study in your thesis and paper
Hi @hosseinfani 
I have written the code to plot the #teams_over_years_dist but there are too many years (ticks) on the x-axis and too many values (ticks) on the y-axis, making the plot look bad. I am trying to figure out how to fix this issue. Also, this has been done using the teams.pkl and not the sparse matrix. I still have not found a way to use the sparse matrix to obtain time related distributions.

**Comment by hosseinfani** [comment]
@VaghehDashti 
thanks. can we have a look at the image here?
we have y2i and i2y index that includes the starting and ending index of rows form teams inside the teamsvec.pkl. I think we can render the distribution using that.

**Comment by VaghehDashti** [comment]
@hosseinfani 
dblp:
[dblp_time_dist.pdf](https://github.com/fani-lab/OpeNTF/files/9737985/dblp_time_dist.pdf)
imdb:
[imdb_time_dist.pdf](https://github.com/fani-lab/OpeNTF/files/9737986/imdb_time_dist.pdf)
uspt:
[uspt_time_dist.pdf](https://github.com/fani-lab/OpeNTF/files/9737987/uspt_time_dist.pdf)

Yes, that's a good idea. I will look into it.

**Comment by hosseinfani** [comment]
@VaghehDashti 
x is the year, but what is y? the number of teams?
If so, dblp and uspt simply say that the number grows within the year, but imdb not.
What important here is the distribution of member-skills in time. That is some skills become depricated while others become trendy in recent time. And members shift their skills within time. 

We have to find a way to show such as thing. Something like:
![image](https://user-images.githubusercontent.com/8619934/194687480-c7562810-dbbe-4af7-8eb5-04b97eb8d7a3.png)

**Comment by cdcorro** [comment]
hi @hosseinfani,

I'm currently trying to adjust the functions that I created to fit in the get_stats() function. However, since my functions mainly use the 'i2y', '12c', 'i2s', etc. data rather than **teamvecs**, I have some questions regarding that function's parameters. 

1.  is teamsvecs a dataframe (from pandas)? if so what are the column names?
2.  does **teamsvecs** also include the year for each team?
3.  is there a way I can get a visualization of what **teamsvecs** is using the toy data?

If I could get an answer to these it would be a great help.

**Comment by hosseinfani** [comment]
Hi @cdcorro 
1. No, it's a dictionary of lil_matrix, that is, {'id': lil_matrix, 'skill': lil_matrix, 'member': lil_matrix}. The file can be loaded by pickle.load()

2. No. There is y2i that is list of tuples (row_index, year) that show what is the starting row_index for a year.
e.g., [(0, 1990),(13, 1991), (25, 1992)...] which means that the rows (teams) from rows 0 till 13-1 is for year 1990, and from 13 till 25-1 is for year 1992, ...

3. Not sure I understood the item. By visualization, do you mean to render it as a dense matrix to see the cells?

**Comment by cdcorro** [comment]
@hosseinfani 

oh okay, that clarifies some things for me. just a few follow up questions:

regarding question 2, get_stats requires these parameters (cls, teamsvecs, output, cache=True, plot=False, plot_title=None). From these parameters is it possible to get the y2i? or do I need to add it as a new parameter to the function?

as for question 3, yes is there a way to render the matrix using the toy data?

**Comment by hosseinfani** [comment]
@cdcorro 

oh, correct, you need to add an arg to pass  the index file

there must be a way. I think matplotlib does that. I've not done that in python tho but rememebr you have first make the sparse matric to dense.

**Comment by VaghehDashti** [comment]
Hi @hosseinfani,
Here are the new plots with proper x and y ticks:
dblp:
[dblp_time_distribution.pdf](https://github.com/fani-lab/OpeNTF/files/9760545/dblp_time_distribution.pdf)
imdb:
[imdb_time_distribution.pdf](https://github.com/fani-lab/OpeNTF/files/9760548/imdb_time_distribution.pdf)
uspt:
[uspt_time_distribution.pdf](https://github.com/fani-lab/OpeNTF/files/9760549/uspt_time_distribution.pdf)

**Comment by cdcorro** [comment]
Hi @hosseinfani,

I was able to move my temporal stats into the get_stats() function. I added the parameter "i2y" to be the array of index to year. The only issue I'm facing is that I wanted to run this get_stats() function to test that everything runs smoothly, but I'm having a hard time finding where the get_stats() function gets called. Also, I was wondering if there was a command to run the stats since I was running the command "python main.py -data data/raw/dblp/toy.dblp.v12.json -domain dblp -model fnn bnn" and no stats were generated to my knowledge.

TLDR;
1. Where is the get_stats() function called? (aka. which files or how do I find the files?)
2. How do I run the stats to test everything is working in order?

**Comment by cdcorro** [comment]
Thank you @hosseinfani. 

I just have a quick question. I printed out the membervecs (which are teamvecs['members']) and I got this format. I was just wondering if you could help me define what I'm looking at as it would help a lot in the debugging processes I'm doing at the moment :)

![image](https://user-images.githubusercontent.com/53372649/198844876-4218b3cf-2525-4cd5-8db4-d4d2859683d9.png)

**Comment by hosseinfani** [comment]
@cdcorro 
the data structure is sparse matrix, only keep the nonzero values.
here the rows show the team index, the column shows the member index, and the value ahows the membership.

(x,y)=1 means that team x has member y

**Comment by cdcorro** [comment]
@hosseinfani 

I was able to finish and test the first iteration of the temporal stats using the toy data. I was trying to make a push request, but it seems that I have not been given permission to push to the repository:

![image](https://user-images.githubusercontent.com/53372649/200634829-740b29df-23a9-4064-ac25-c98a54ce94b3.png)

**Comment by hosseinfani** [comment]
@cdcorro 
strange! can you make pr?

**Comment by cdcorro** [comment]
@hosseinfani 

I am unable to push any of the code I added. I just get the permission error message as stated above.

**Comment by hosseinfani** [comment]
@cdcorro pls try now

**Comment by cdcorro** [comment]
@hosseinfani 

Thank you!  I was able to make a PR (https://github.com/fani-lab/OpeNTF/pull/181). I made it with a branch called "Charles-branch". If you have any questions just let me know :)
